flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/QueryOperationVisitor.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations;
+
+import org.apache.flink.annotation.Internal;
+
+/**
+ * Class that implements visitor pattern. It allows type safe logic on top of tree
+ * of {@link QueryOperation}s.
+ */
+@Internal
+public interface QueryOperationVisitor<T> {
+
+	T visitProject(ProjectQueryOperation projection);
+
+	T visitAggregate(AggregateQueryOperation aggregation);
+
+	T visitWindowAggregate(WindowAggregateQueryOperation windowAggregate);
+
+	T visitJoin(JoinQueryOperation join);
+
+	T visitSetOperation(SetQueryOperation setOperation);
+
+	T visitFilter(FilterQueryOperation filter);
+
+	T visitDistinct(DistinctQueryOperation distinct);
+
+	T visitSort(SortQueryOperation sort);
+
+	<U> T visitCalculatedTable(CalculatedQueryOperation<U> calculatedTable);
+
+	T visitCatalogTable(CatalogQueryOperation catalogTable);
+
+	T visitCatalogSink(CatalogSinkOperation catalogSink);
+
+	<U> T visitInlineSink(UnregisteredSinkOperation<U> inlineSink);
```

Comment:
I think this method name should be `visitUnregisteredSink` ?

Reply:
nit: let's simplify the method names to `visit(...)` only. I think this is pretty common.